In [5]:
import sqlite3
from sqlite3 import Error

 
def create_connection(db_file):
    """ create a database connection to the SQLite database
        specified by the db_file
    :param db_file: database file
    :return: Connection object or None
    """
    conn = None
    try:
        conn = sqlite3.connect(db_file)
        print(sqlite3.version)
    except Error as e:
        print(e)
 
    return conn

 
def select_all_tasks(conn):
    """
    Query all rows in the tasks table
    :param conn: the Connection object
    :return:
    """
    cur = conn.cursor()
    
    query1 = """
        SELECT f.name AS facility_name,
       SUM(b.slots) AS usage_by_member
FROM Facilities f
INNER JOIN Bookings b
ON f.facid = b.facid
INNER JOIN Members m
ON b.memid = m.memid
WHERE b.memid != 0
GROUP BY f.name
ORDER BY usage_by_member DESC;
        """
    cur.execute(query1)
 
    rows = cur.fetchall()
 
    for row in rows:
        print(row)


def main():
    database = "sqlite_db_pythonsqlite.db"
 
    # create a database connection
    conn = create_connection(database)
    with conn: 
        print("2. Query all tasks")
        select_all_tasks(conn)
 
 
if __name__ == '__main__':
    main()

2.6.0
2. Query all tasks
('Badminton Court', 1086)
('Tennis Court 1', 957)
('Massage Room 1', 884)
('Tennis Court 2', 882)
('Snooker Table', 860)
('Pool Table', 856)
('Table Tennis', 794)
('Squash Court', 418)
('Massage Room 2', 54)


/* Q10: Produce a list of facilities with a total revenue less than 1000.
The output of facility name and total revenue, sorted by revenue. Remember
that there's a different cost for guests and members! */

2.6.0
2. Query all tasks
('Table Tennis', 180)
('Snooker Table', 240)
('Pool Table', 270)

Here's the code for query1 variable:

SELECT f.name AS facility_name,
       SUM(
           CASE 
               WHEN b.memid = 0 THEN f.guestcost * b.slots
               ELSE f.membercost * b.slots
           END
       ) AS total_revenue
FROM Facilities f
INNER JOIN Bookings b
ON f.facid = b.facid
LEFT JOIN Members m
ON b.memid = m.memid
GROUP BY f.name
HAVING total_revenue < 1000
ORDER BY total_revenue;



/* Q11: Produce a report of members and who recommended them in alphabetic surname,firstname order */
    
    localhost/admin_springboard/Members/		https://frankfletcher.co/springboard_phpmyadmin/index.php?route=/table/sql&db=admin_springboard&table=Members

   Showing rows 0 - 24 (31 total, Query took 0.0016 seconds.)


SELECT CONCAT(m1.surname, ', ', m1.firstname) AS member_name,
       CONCAT(m2.surname, ', ', m2.firstname) AS recommender_name
FROM Members m1
LEFT JOIN Members m2
ON m1.recommendedby = m2.memid
ORDER BY m1.surname, m1.firstname;


member_name	recommender_name	
Bader, Florence	Stibbons, Ponder	
Baker, Anne	Stibbons, Ponder	
Baker, Timothy	Farrell, Jemima	
Boothe, Tim	Rownam, Tim	
Butters, Gerald	Smith, Darren	
Coplin, Joan	Baker, Timothy	
Crumpet, Erica	Smith, Tracy	
Dare, Nancy	Joplette, Janice	
Farrell, David	GUEST, GUEST	
Farrell, Jemima	GUEST, GUEST	
Genting, Matthew	Butters, Gerald	
GUEST, GUEST	GUEST, GUEST	
Hunt, John	Purview, Millicent	
Jones, David	Joplette, Janice	
Jones, Douglas	Jones, David	
Joplette, Janice	Smith, Darren	
Mackenzie, Anna	Smith, Darren	
Owen, Charles	Smith, Darren	
Pinker, David	Farrell, Jemima	
Purview, Millicent	Smith, Tracy	
Rownam, Tim	GUEST, GUEST	
Rumney, Henrietta	Genting, Matthew	
Sarwin, Ramnaresh	Bader, Florence	
Smith, Darren	GUEST, GUEST	
Smith, Darren	GUEST, GUEST	


For some reason, CONCAT isn't recognized in this Jupyter notebook, but it works just fine in the phpMyAdmin portal. So, I've copied the results from that portal here into this notebook.



/* Q12: Find the facilities with their usage by member, but not guests */

2.6.0
2. Query all tasks
('Badminton Court', 1086)
('Tennis Court 1', 957)
('Massage Room 1', 884)
('Tennis Court 2', 882)
('Snooker Table', 860)
('Pool Table', 856)
('Table Tennis', 794)
('Squash Court', 418)
('Massage Room 2', 54)


Here's the query1 variable code:

SELECT f.name AS facility_name,
       SUM(b.slots) AS usage_by_member
FROM Facilities f
INNER JOIN Bookings b
ON f.facid = b.facid
INNER JOIN Members m
ON b.memid = m.memid
WHERE b.memid != 0
GROUP BY f.name
ORDER BY usage_by_member DESC;



/* Q13: Find the facilities usage by month, but not guests */
    
 localhost/admin_springboard/Facilities/		https://frankfletcher.co/springboard_phpmyadmin/index.php?route=/table/sql&db=admin_springboard&table=Facilities

   Showing rows 0 - 24 (27 total, Query took 0.0126 seconds.)


SELECT f.name AS facility_name,
       YEAR(b.starttime) AS year,
       MONTH(b.starttime) AS month,
       SUM(b.slots) AS usage_by_member
FROM Facilities f
INNER JOIN Bookings b
ON f.facid = b.facid
INNER JOIN Members m
ON b.memid = m.memid
WHERE b.memid != 0
GROUP BY f.name, year, month
ORDER BY year, month, usage_by_member DESC;


facility_name	year   	month   	usage_by_member   	
Tennis Court 1	2012	7	201	
Massage Room 1	2012	7	166	
Badminton Court	2012	7	165	
Snooker Table	2012	7	140	
Tennis Court 2	2012	7	123	
Pool Table	2012	7	110	
Table Tennis	2012	7	98	
Squash Court	2012	7	50	
Massage Room 2	2012	7	8	
Badminton Court	2012	8	414	
Tennis Court 2	2012	8	345	
Tennis Court 1	2012	8	339	
Massage Room 1	2012	8	316	
Snooker Table	2012	8	316	
Pool Table	2012	8	303	
Table Tennis	2012	8	296	
Squash Court	2012	8	184	
Massage Room 2	2012	8	18	
Badminton Court	2012	9	507	
Pool Table	2012	9	443	
Tennis Court 1	2012	9	417	
Tennis Court 2	2012	9	414	
Snooker Table	2012	9	404	
Massage Room 1	2012	9	402	
Table Tennis	2012	9	400	
   

This code works in phpMyAdmin, but not in this Jupyter notebook.
So, I ran the query in the portal and copied it into this notebook.